### ISSUE TITLE
Inference is producing better results when the model is in ".train()" mode?

### ISSUE THREAD

**Issue by Ammartauqir** [xinge008/Cylinder3D#54]
Thanks for the great work!
I wrote a script in order to infer the semantic labels for the test dataset. I observed that after I loaded the trained model and give test data as an input to the model, the inferred results are coming out to be better when the loaded model is in 'model.train()' mode while the performance deteriorates when I put the model into 'model.eval()' mode for inference, as per the official documentation recommends. I might be missing something. What could be the potential reason for that?

Thank you 
Best regards

**Comment by xinge008** [comment]
Can you give details of these two results, i.e., result on each category?

How do you perform the inference with 'model.train()'? Just remove model.eval() and others remain unchange?

**Comment by Ammartauqir** [comment]
After I have loaded the trained model, I use the following chunk of code in order to generate labels and save them in a directory.  The problem occurs if uncomment 'model.eval()' in the script: 

model = model_builder.build(model_config)
model.load_state_dict(torch.load(model_load_path,map_location='cuda:0')  model.to(pytorch_device)
    train_dataset_loader, val_dataset_loader, test_dataset_loader = data_builder.build(dataset_config,train_dataloader_config, val_dataloader_config, grid_size=grid_size)

####
    # model.eval()   <<<<< Causing Problem
####
    with torch.no_grad():
        count=0
        hist_list = []
        for i_iter_val, (_, vox_label, grid, pt_labs , pt_fea) in tqdm(enumerate(test_dataset_loader), total=len(test_dataset_loader.dataset.point_cloud_dataset.im_idx)):
            pass
            val_pt_fea_ten = [torch.from_numpy(i).type(torch.FloatTensor).to(pytorch_device) for i in
                              pt_fea]
            val_grid_ten = [torch.from_numpy(i).to(pytorch_device) for i in grid]
            # val_label_tensor = val_vox_label.type(torch.LongTensor).to(pytorch_device)

            predict_labels_raw = model(val_pt_fea_ten, val_grid_ten, val_batch_size)
            predict_argmax = predict_labels_raw[0].argmax(dim=0)
            predict_labels = predict_argmax.cpu().detach().numpy()
            predict_labels_serialized=[]
            for n_index, grid_index in enumerate(grid[0]):
                predict_labels_serialized.append(predict_labels[grid_index[0]][grid_index[1]][grid_index[2]])
            predict_labels_serialized = np.array(predict_labels_serialized, dtype=np.int32)
            for index, label in enumerate(predict_labels_serialized):
                predict_labels_serialized[index] = SemKITTI_learningmap_inv[label]
            # print(predict_labels_serialized.size)
            sample_name = test_dataset_loader.dataset.point_cloud_dataset.im_idx[i_iter_val][-10:-4]
            sequence_num = test_dataset_loader.dataset.point_cloud_dataset.im_idx[i_iter_val][-22:-20]

            path_to_seq_folder = path_to_save_pridicted_labels + '/' + str(sequence_num)
            if not os.path.exists(path_to_seq_folder):
                os.makedirs(path_to_seq_folder)


            # print(sample_name)
            # print(sequence_num)
            xyz_point_cloud = pt_fea[0][:,5:9]
            temp = np.array([xyz_point_cloud[:,2]])
            xyz_point_cloud[:,2] = xyz_point_cloud[:,0]
            xyz_point_cloud[:, 0] = temp
            xyz_point_cloud = np.array(xyz_point_cloud, dtype=np.float32)
            # xyz_point_cloud = np.vstack(([xy_point_cloud],[z_point_cloud]))

            predict_labels_serialized.tofile(path_to_seq_folder+'/' + sample_name + '.label')
            count+=1

**Comment by Ammartauqir** [comment]
Following are the per class results of the labels generated from the above script.

![image](https://user-images.githubusercontent.com/16537287/115363381-b528a800-a1c2-11eb-9015-5aff90e859c6.png)

**Comment by xinge008** [comment]
It seems that you use multi-scan setting? I saw some categories, including moving-person, moving-bicylist, that do not belong to single-scan semantic segmentation.

**Comment by Ammartauqir** [comment]
Thank you for your response. Yea, this is a Multi Scan evaluation.

**Comment by xinge008** [comment]
I do not release the pretrained model for multi-scan semantic segmentation. If you load the single-scan pretrained model, its performance would drop significantly.

**Comment by Ammartauqir** [comment]
Sorry that was a mistake. I have performed evaluation with the single scan. Following are the results.
![image](https://user-images.githubusercontent.com/16537287/115882672-457a1d80-a44d-11eb-83e4-2404d95ac2d2.png)

**Comment by xinge008** [comment]
Maybe the version of pytorch and some other dependencies make the difference; You can refer to this issue https://github.com/xinge008/Cylinder3D/issues/30

**Comment by awethaileslassie** [comment]
@Ammartauqir @xinge008 
May I ask how you were able to train the model in multiscan Semantic-Segmentation?
I have tried to but there is some problem with the input dimension.
